### PR TITLE
Improve logging

### PR DESCRIPTION
**Pull request by relu91**
This PR implements and fixes what described in #229. I did not modify log messages generated inside examples and tests as I think them as different from the one created inside the library itself.

### CODE REVIEW

**Comment by egekorkan** [comment]
Just for the clarifyin something for me, when you use console.debug, it will log only in the debug level and below right? Since this is a modification that some might be relying on ( I do for some small device implementations), it should be explained somewhere how to enable logging in general. The cli config does this already but maybe already in the readme? I say this since everything was logged even if the developer was not specifying anything, it breaks their expectation.

File: README.md [hunk 1]
```diff
@@ -274,6 +274,12 @@ Note: More mediaTyes can be easily added by implementing `ContentCodec` interfac
 
 We used to have a node-wot-logger package to allow fine-grained logging (by means of Winston). It turned out though that depending on the actual use-case other logging libraries might be better suited. Hence we do not want to prescribe which logging library to use. Having said that, we use console statements which can be easily overriden to use the prefered logging library if needed (see [here](https://gist.github.com/spmason/1670196)).
 
+The logs in the library follows those best practice rules (see [here](https://github.com/eclipse/thingweb.node-wot/issues/229)):
```
Comment (egekorkan) [inline README.md:277]: My comment at https://github.com/eclipse/thingweb.node-wot/pull/250#issuecomment-644993618 can be added somewhere around this paragraph

File: packages/binding-http/src/http-client.ts [hunk 2]
```diff
@@ -90,16 +90,16 @@ export default class HttpClient implements ProtocolClient {
 
   public async readResource(form: HttpForm): Promise<Content> {
     const request = await this.generateFetchRequest(form, "GET")
-    console.info(`HttpClient (readResource) sending ${request.method} to ${request.url}`);
+    console.debug("[binding-http]",`HttpClient (readResource) sending ${request.method} to ${request.url}`);
```
Comment (egekorkan) [inline packages/binding-http/src/http-client.ts:93]: mapping info to debug, is it intended?

**Comment by egekorkan** [comment]
So my general comment is to document the proposed mapping. Please change if the following is not correct:
- log -> debug
- info -> debug
- error -> error
- warn -> warn

**Comment by egekorkan** [comment]
It would be also good to differentiate between cli and cli-default-servient since both can be used at the same time whereas in the other packages they are either client or server and are less likely to be used together

**Comment by egekorkan** [comment]
Please don't take my comments badly, I think the improvement is more than a slight improvement! Thank you @relu91 :)

**Comment by relu91** [comment]
Thank you Ege for your comments, here my thoughts. 

>Just for the clarifyin something for me, when you use console.debug, it will log only in the debug level and below right? Since this is a modification that some might be relying on ( I do for some small device implementations), it should be explained somewhere how to enable logging in general. The cli config does this already but maybe already in the readme? I say this since everything was logged even if the developer was not specifying anything, it breaks their expectation.

In my understanding Node console does not really have the concept of *log levels*. In fact, most of the functions supported by console are just aliases for `console.log` and `console.error` (see [here](https://nodejs.org/docs/latest-v10.x/api/console.html#console_console_debug_data_args) for example). So when we use `console.debug` we are just using `console.log`. So why bother? Well using that function we can emulate *log levels*. For example, consider the following as the default level order:

0. Off
1. Error
2. Warning
3. Info
4. Debug

Setting a level the log to 2 means that `console.info = () => {}` and `console.debug=() =>{}` . This PR is not about adding this functionality. @fatadel is working on this, because I see it as a different feature (i.e. implement log levels). 

>log -> debug
info -> debug
error -> error
warn -> warn

I was unsure to map info to debug. However, most of the info messages seemed more suited in the debug messages. We can discuss which one may make more sense as an info message even if I think that the lesser level we use the easier is to manage them. 

On the other hand about `console.log` I think that most of the users just do not care about a complex logging system and in the end just use `console.log`. Using this function in node-wot will make it harder for such users to shut down the logs completely (i.e. we cannot do  `console.log = () => {}` because it shuts down also application messages).  

So the rule is not a mapping, it is more do not use log at all and think twice before use `info`. 

>It would be also good to differentiate between cli and cli-default-servient since both can be used at the same time whereas in the other packages they are either client or server and are less likely to be used together.
Similarly for core which has other subpackages inside

This could improve even more the detection of the log messages. I thought a pattern like the following: `[packageName:ScriptName] Some message`. What do you think? too long?

**Comment by egekorkan** [comment]
- Thanks for the clarification regarding upcoming PR by @fatadel :)
- Regarding not having simple console.log, it also makes a lot of sense.
- Agree on having less levels. In the end how should I make a decision whether I want warn or info :)
- `[packageName:ScriptName]` makes sense, e.g. `[core-codec/json]`. I think this would be needed only for the core packages since the protocol bindings differentiate themselves by being Consumer or Server anyways but it wouldn't hurt to add the suffix there as well :)

**Comment by relu91** [comment]
> [packageName:ScriptName] makes sense, e.g. [core-codec/json]. I think this would be needed only for the core packages since the protocol bindings differentiate themselves by being Consumer or Server anyways but it wouldn't hurt to add the suffix there as well :) 

Ok, I'll change the PR to draft while I'm adding that change.

**Comment by danielpeintner** [comment]
@relu91 Thank your very much for all the work.

I looked at the PR and it looks very good.

Some comments/questions below.

### I found `console.info` in one file

I guess we might find more but we can tackle them once we encounter them.

packages\binding-coap\src\coaps-client.ts, line 149 --> console.info(`HttpClient overriding client-side proxy with security proxyURI '${security.proxyURI}`);

### Id/Ids in one file per package?
One question I would like to raise is whether it makes sense to have a *single* place in each package for the identifier like "[binding-http]". This would make it simple to change it in the future (IF we ever want to do so).
In some cases (other than binding) we have multiple identifiers like "[core/helpers]", "[core/exposed-thing]", .. but even there it might make sense to have all const values in a single place?

Does that make sense or is it not worth it?
I am myself unclear what is best...

### Consistent identifier
Question: in in all bindings we use the package name except "[binding-ws]" where it might be "[binding-websockets]" to be consistent?


### Another nit-picking comment is about formatting.

When using the VisualStudio code formatting it adds spaces between , et cetera.

For example
`console.debug("[binding-http]",`HttpClient (readResource) sending ${request.method} to ${request.url}`);`
becomes
`console.debug("[binding-http]", `HttpClient (readResource) sending ${request.method} to ${request.url}`);`

Does it make sense to do this clean-up also in this PR? Or shall we do that formatting clean-up in another PR we plan to make anyway...

### Merge conflict 
We also need to fix the merge conflict that was raised by a recent PR merge...

**Comment by relu91** [comment]
Thank you @danielpeintner for the great review! here they are my thoughts:
> I found console.info in one file
I guess we might find more but we can tackle them once we encounter them.

packages\binding-coap\src\coaps-client.ts, line 149 --> console.info(HttpClient overriding client-side proxy with security proxyURI '${security.proxyURI});

Nice catch! it escaped my review process thank you for point it out. 

> Does that make sense or is it not worth it?
I am myself unclear what is best...

It does make sense. However, I am afraid that going that path will lead us to have a full-blown logging framework... with this PR, I wanted to make a micro-step toward better log management. I mean, if we gonna use "global/local" `const` tags variables why not make a dedicated logger class for each package --> why don't we factorize it a stand-alone package? --> now we have a logging framework. As you said I am not certain that we do need/not need this. 

>Question: in in all bindings we use the package name except "[binding-ws]" where it might be "[binding-websockets]" to be consistent?

I agree.

> Another nit-picking comment is about formatting.

I am not sure if I can do this without formatting the entire file. One question: this means that my vscode instance is not using the `.editorconfig` rules?

> We also need to fix the merge conflict that was raised by a recent PR merge...

I'll do it 👍 

Changing back to draft while I'll make the changes requested above

**Comment by relu91** [comment]
Just find out that `packages\binding-coap\src\coaps-client.ts, line 149` is a commented line. I'll leave the comment as it is.

**Comment by danielpeintner** [comment]
> Just find out that `packages\binding-coap\src\coaps-client.ts, line 149` is a commented line. I'll leave the comment as it is.

Sorry, my bad!

> > Another nit-picking comment is about formatting.
> 
> I am not sure if I can do this without formatting the entire file. One question: this means that my vscode instance is not using the `.editorconfig` rules?

I have to explicitly say "Right mouse click" --> "Format selection" ...
It is *not* done automatically.

> > Does that make sense or is it not worth it?
> > I am myself unclear what is best...
> 
> It does make sense. However, I am afraid that going that path will lead us to have a full-blown logging framework... with this PR, I wanted to make a micro-step toward better log management. I mean, if we gonna use "global/local" `const` tags variables why not make a dedicated logger class for each package --> why don't we factorize it a stand-alone package? --> now we have a logging framework. As you said I am not certain that we do need/not need this.

Maybe you are right and we better stick with the *simple* approach.
Any other opinions?

If not, I am OK with merging!

**Pull request merged**